sys/include/net/nanocoap.h

```diff
@@ -1500,6 +1500,61 @@ static inline ssize_t coap_opt_add_uri_path_buffer(coap_pkt_t *pkt,
     return coap_opt_add_chars(pkt, COAP_OPT_URI_PATH, path, path_len, '/');
 }
 
+/**
+ * @brief   Adds an entity-tag (etag) value to the coap pkt.
+ *
+ * @param[in,out] pkt         Packet being built
+ * @param[in]     etag        Etag value of the packet
+ * @param[in]     len         Length of @p etag in bytes
+ *
+ * @return        number of bytes written to pkt buffer
+ * @return        -EINVAL if invalid etag size is used for @p len
+ * @return        -ENOSPC if no available options or pkt full
+ */
+static inline ssize_t coap_opt_add_etag(coap_pkt_t *pkt, const uint8_t *etag, size_t len)
+{
+    if (len == 0 || len > 8) {
+        return -EINVAL;
+    }
+    return coap_opt_add_opaque(pkt, COAP_OPT_ETAG, etag, len);
+}
+
+/**
+ * @brief   Adds an empty dummy entity-tag (etag) value to the coap pkt.
+ *
+ * This function adds an all-zero etag value to the packet as placeholder for later updating
+ *
+ * @param[in,out] pkt         Packet being built
+ * @param[in]     len         Length of etag space to reserve in bytes
+ *
+ * @return        number of bytes written to pkt buffer
+ * @return        -EINVAL if invalid etag size is used for @p len
+ * @return        -ENOSPC if no available options or pkt full
+ */
+static inline ssize_t coap_opt_add_etag_dummy(coap_pkt_t *pkt, size_t len)
+{
+    const uint8_t zeros[8] = { 0 };
+    return coap_opt_add_etag(pkt, zeros, len);
+}
+
+
+/**
+ * @brief   replaces an entity-tag (etag) value with a new one in a coap pkt.
+ *
+ * @note    The etag option must have been added first via @ref coap_opt_add_etag or
+ *          @ref coap_opt_add_etag_dummy
+ * @note    The @p len supplied to this call must match the length supplied to the call to add the
+ *          etag option
```

Comment:
This makes this function basically unusable for cache re-validation at the client. Unless the user somehow looks up the cache entry and fetches the exact length the server supplied. But for that they need a somewhat complete request (ignoring questions which are not accounted for in cache key validation), which causes a bootstrapping problem. Its why I did the complicated dance of adding a full-length `COAP_ETAG_LENGTH_MAX` dummy ETag in 728c7d60889 first only to use memmove later to reduce the size if necessary.

Reply:
I'm not sure what to do here. For me the use case here was limited to known etag lengths for generating server responses.
 I think it would significantly increase complexity of replacing the etag if you want to relax this condition. It would also mean informing the user of the function with how many bytes were removed from the packet.

At this point it might be easier to have a generic `coap_opt_insert_opaque()`

Reply:
Such applications can still use coap_opt_add_etag. When building a response from a cache or from a response received from the network, I'd expect that building the response only happens when the input is fully available, and thus the ETag can be calculated. Only when the ETag is calculated on the fly while populating the payload, then the two-part dummy steps are needed. (But to be fair, I don't quite understand what the nanocoap cache ETag tricks do in your linked commit in the first place).

Reply:
> (But to be fair, I don't quite understand what the nanocoap cache ETag tricks do in your linked commit in the first place).

It assumes that the client user does not need to know about the cache. So to allow for cache validation (i.e. get the ETag from the cache) it adds first a dummy ETag option of full length on request initialization and then populates it once it has a cache hit before send. Since that ETag comes from the server, it can be of any length, so the dummy option needs to be resized (after cache key calculation, so basically after the whole message is already constructed).

Reply:
I think I managed to rewrite this to allow for arbitrary etag sizes in the replace code, as long as the initial etag was larger than the new etag. The added complexity is not as much as I initially expected, but I'm rather curious whether I'm covering all edge cases. I'm also not very happy with the footgun that might be the return value

The example code does seem to still work though.